Simplenote/NoticeController.swift

```diff
@@ -0,0 +1,89 @@
+import UIKit
+
+class NoticeController {
+    static let shared = NoticeController()
+
+    private var notices: [Notice] = []
+    private var current: Notice?
+    private let noticePresenter = NoticePresenter()
+
+    private var activeViewIsBeingTouched: Bool = false
+
+    private init() { }
+
+    func setupNoticeController() {
+        noticePresenter.startListeningToKeyboardNotifications()
+    }
+
+    func present(_ notice: Notice) {
+        if noticePresenter.isPresenting {
+            appendToQueueIfNew(notice)
+            return
+        }
+
+        current = notice
+        let noticeView = makeNoticeView(from: notice)
+
+        noticePresenter.presentNoticeView(noticeView) {
+            if !self.activeViewIsBeingTouched {
+                self.dismiss(noticeView)
+            }
+        }
+    }
+
+    private func dismiss(_ noticeView: NoticeView) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + Times.tapDelay) {
+            self.noticePresenter.dismissNotification {
+                self.current = nil
+                if !self.notices.isEmpty {
+                    self.present(self.notices.removeFirst())
+                }
+            }
+        }
+    }
+
+    private func makeNoticeView(from notice: Notice) -> NoticeView {
+        let noticeView: NoticeView = NoticeView.instantiateFromNib()
+        noticeView.message = notice.message
+        noticeView.actionTitle = notice.action?.title
+        noticeView.handler = notice.action?.handler
+        noticeView.delegate = self
+
+        return noticeView
+    }
+
+    /// Confirms if a notice is already contained in notices queue. Appends to queue if new
+    ///
+    private func appendToQueueIfNew(_ notice: Notice) {
+        if notices.contains(notice) {
+            return
+        }
+
+        if let current = current {
+            if notice == current {
+                return
+            }
+        }
+
+        notices.append(notice)
+    }
+}
+
+extension NoticeController: NoticePresentingDelegate {
+    func noticePressBegan() {
+        activeViewIsBeingTouched = true
+    }
+
+    func noticePressEnded() {
+        activeViewIsBeingTouched = false
+
+        guard let noticeView = noticePresenter.noticeView else {
+            return
+        }
+        dismiss(noticeView)
+    }
+}
```

Comment:
My thought here was that the controller acts as the triggering mechanism for changes to the notices, and the presenter handles the animations.  So the animation should certainly be in the presenter, I moved that.

I have left the tap responding in the controller, mostly because I am retriggering the dismiss from the controller once the press is finished.  If I trigger the animation dismissal from the presenter, I don't have a way of letting the controller know that the dismissal is complete currently.  By having the controller handle the long press, it can re trigger it's dismissal method so the completion handler can be resolved to the controller when animation is complete